lib/Runtime/Library/JSON.cpp

```diff
@@ -682,56 +681,90 @@ namespace JSON
                 Js::JavascriptStaticEnumerator enumerator;
                 if (object->GetEnumerator(&enumerator, EnumeratorFlags::SnapShotSemantics | EnumeratorFlags::EphemeralReference, scriptContext))
                 {
-                    Js::DynamicObject * dynamicObject = ((Js::DynamicObject*)object);
-                    if (ReplacerFunction != replacerType || !dynamicObject->HasObjectArray())
+                    bool isDynamicObject = Js::DynamicObject::Is(object);
+                    Js::DynamicObject * dynamicObject = nullptr;
+                    bool hasObjectArray = false;
+                    if (isDynamicObject)
                     {
-                        uint32 propertyCount = dynamicObject->GetPropertyCount();
+                        dynamicObject = Js::DynamicObject::FromVar(object);
+                        if (dynamicObject->HasObjectArray())
+                        {
+                            hasObjectArray = dynamicObject->HasObjectArray();
+                        }
+                    }
 
-                        result = Js::ConcatStringBuilder::New(this->scriptContext, propertyCount);
-                        Js::DynamicTypeHandler * typeHandler = dynamicObject->GetTypeHandler();
+                    if (ReplacerFunction != replacerType || !hasObjectArray)
```

Comment:
Could you please add the `!isDynamicObject` condition here? I know it is redundant since that requires that `!hasObjectArray`, but that's not immediately clear and I spent a while trying to work out what combination of this conditional and the following conditional could happen.

Reply:
`!hasObjectArray` != `!isDynamicObject`

Reply:
Condition here is.. Everything but `ES5 && ReplacerFunction == replacerType. && object->HasObjectArray`

Reply:
Because, when there is a replacer function && we need to do an enumerate-loop (because of objectArray), we will need to store the existing properties. That's what the `else` of `this if` does

Reply:
Ah; I was mis-reading the indentation depth. Let me take a closer look and make sure I understand.

Reply:
Actually, ruling out `!dynamicObject && replacerType != replacerFunction` was a good idea